### DISCUSSION TITLE
Intended solution for harvesting two projects with shared files

### DISCUSSION THREAD

**Discussion by Toby222** [wixtoolset/issues#8877] [Questions]
I'm trying to create a setup where one of the features requires two projects (a backend executable and a database migration tool)
The two projects share a bunch of libraries, so if I try to harvest both of their output folders into the same install folder, I logically get WIX0091 errors for duplicate files. Is there any way to deduplicate files like this?
Or what would be the simplest way to achieve this without having to resort to manually specifying the library files?

**Comment by nirbar** [comment]
If you're absolutely sure that both projects use the same set of library versions, then you can build both into the same output folder and harvest them once
If they might be using different versions of the libraries then you better install them into different folders

**Reply by Toby222** [reply]
The libraries are our own first party projects, so yea they're using the same, but building into the same folder seems counter-intuitive as well

**Reply by nirbar** [reply]
If you're deploying to the same folder, why is building to the same folder counterintuitive?

Another way would be harvesting a parent folder of the two build folders with a xslt to merge both folders into a single install folder

**Reply by Toby222** [reply]
Counterintuitive since I've only ever seen projects output to their own directory, and it feels like stuff like the detection if a project is up-to-date would break if I had two projects build in one folder

Plus I've got another project that actually gets installed to several other Features ^^;
(That one I'm hoping I can replace with a custom action or something in the future, something to properly set the App.config values after install)

**Reply by chrpai** [reply]
I think you could benefit from a probono 1 hour discussion on set theory as it relates to componetization of files, the benefits of static authoring vs dynamic harvesting and a FOSS tool that I have that helps with all of this.

**Comment by robmen** [comment]
At FireGiant, we call these "diamond dependencies", and our [Advanced Harvesting](https://www.firegiant.com/docs/fg-wix/harvesting/), which is part of the "HeatWave Build Tools" supports them. Today, you get those tools as part of [FireGiant support contract](https://www.firegiant.com/services/support-programs/).

**Reply by Toby222** [reply]
From what that post says, I'm not sure it'd solve the issue of duplicate files?
My problem is harvesting (or installing) two projects with overlapping dependencies into one folder without unnecessarily duplicating the libraries.
Either way I take it this isn't possible with just the open source wix tooling

**Reply by robmen** [reply]
The documentation for Advanced Harvesting doesn't explicitly call or the handling of diamonds, it probably should.

No, it is not part of the Open Source project. It's one of the features we offer from FireGiant. Is that an issue?

**Reply by Toby222** [reply]
Absolutely not in a position to get enterprise tools for this :^)
I'll figure something out, thanks anyway

**Comment by chrpai** [comment]
In a wix v4/5 world you could simply publish the two apps to two different directories and then use Files to dynamically harvest them.  They will author duplicate components (but different directories)  and while this will take up more space on the target machine it won't make the MSI any bigger due to smart cabbing.

I went deep down this road once upon a time and decided to pass on harvesting.  Free consultations are available to discuss how I'd handle it if you'd like.

**Reply by Toby222** [reply]
That's exactly the solution I wanted to avoid :p

**Comment by bevanweiss** [comment]
It sounds like you potentially want to have THREE projects.
One for your common dependencies, one for your backend, and one for the migration.

Then you can have three component groups:
1. Common dependencies
2. Backend (excluding common dependencies)
3. Migration (excluding common dependencies)

Since you wouldn't then have duplicates, it shouldn't be a problem.
You could potentially even factor these three things out into their own MSIs, and then have a bundle which installs them.  This would allow you to potentially upgrade dependencies without having to also update the Backend/Migration installations.

If the backend and the migration don't actually communicate, then I'd suggest not artificially linking the dependencies.
Have them install to separate folder locations, and shipping their own independent dependencies.

**Reply by Toby222** [reply]
I wouldn't have duplicate components, but still have duplicate files installed, which I'd rather want to avoid.
Sadly this seems like the best way to go about it though, I'll see about implementing it

**Reply by robmen** [reply]
Personally, I'd argue the best way is to pick a tool that handles this scenario well. But you can trade time instead of money and go with an alternative solution.

**Reply by Toby222** [reply]
I have significantly more time than money :grin:

**Reply by robmen** [reply]
Ahh, is this for a personal project or an OSS project you are working on?

**Reply by Toby222** [reply]
> Ahh, is this for a personal project or an OSS project you are working on?

No it's for my job, but I couldn't convince anyone to pay for anything cuz I'm just a lowly programmer :p

**Reply by robmen** [reply]
Oh, that's interesting. So, if the WiX Toolset wasn't free, would you be using it?

**Reply by Toby222** [reply]
I'm obviously not the one in charge of deciding that, but most likely not, no.
We have another team that uses something else for setups, so I presume we'd have ended up using that as well

**Reply by robmen** [reply]
Interesting. So, why aren't you using that other option?

**Reply by Toby222** [reply]
Legacy reasons, I believe? It was already the thing we were using before I was hired, and since then we've never had time or reason to rebuild the setup from scratch in another program

**Reply by robmen** [reply]
Interesting. So why did you start using WiX?

**Reply by Toby222** [reply]
dunno, it was already the thing we were using before I was hired :v
("it" being WiX)
